src/Agent.Sdk/LogPlugin.cs

```diff
@@ -0,0 +1,534 @@
+            string typeName = plugin.GetType().FullName;
+            var context = _pluginContexts[typeName];
+
+            bool initialized = false;
+            try
+            {
+                initialized = await plugin.InitializeAsync(context);
+            }
+            catch (Exception ex)
+            {
+                errors.Add($"Fail to initialize: {ex.Message}.");
+                context.Trace(ex.ToString());
+            }
+            finally
+            {
+                if (!initialized)
+                {
+                    context.Trace("Skip process outputs base on plugin initialize result.");
+                    _shortCircuited[typeName].TrySetResult(0);
+                }
+            }
+
+            using (var registration = token.Register(() =>
+                                      {
+                                          var depth = _outputQueue[typeName].Count;
+                                          if (depth > 0)
+                                          {
+                                              context.Output($"Waiting for log plugin to finish, pending process {depth} log lines.");
+                                          }
+                                      }))
+            {
+                while (!_shortCircuited[typeName].Task.IsCompleted && !token.IsCancellationRequested)
+                {
+                    while (!_shortCircuited[typeName].Task.IsCompleted && _outputQueue[typeName].TryDequeue(out string line))
+                    {
+                        try
+                        {
+                            var id = line.Substring(0, line.IndexOf(":"));
+                            var message = line.Substring(line.IndexOf(":") + 1);
+                            var processLineTask = plugin.ProcessLineAsync(context, _steps[id], message);
+                            var completedTask = await Task.WhenAny(_shortCircuited[typeName].Task, processLineTask);
+                            if (completedTask == processLineTask)
+                            {
+                                await processLineTask;
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            // ignore exception
+                            // only trace the first 10 errors.
+                            if (errors.Count < 10)
+                            {
+                                errors.Add(ex.ToString());
+                            }
+                        }
+                    }
+
+                    // back-off before pull output queue again.
+                    await Task.Delay(500);
+                }
+            }
+
+            // process all remaining outputs
+            context.Trace("Process remaining outputs after job finished.");
+            while (!_shortCircuited[typeName].Task.IsCompleted && _outputQueue[typeName].TryDequeue(out string line))
```

Comment:
Consider extracting this part of the code into a function. Then, you'll be able to de-dupe the section of identical code above.

Reply:
will do.